### ISSUE TITLE
Feature request: top-down dependency injection

### ISSUE THREAD

**Issue by ShortRoundDev** [L2jLiga/fastify-decorators#835]
The `@Service` decorator is only called if the implementation of the service is explicitly called. If this typename is known to the constructor, it kind of defeats the point of DI as the controller now has a dependency at least on the name and path of the dependency.

It would be helpful to have a top-down DI initialization where the token is set at fastify initialization.

example use case:

```ts
//IBlogService.ts
import { BlogPost } from './models/BlogPost';

export interface IBlogService
{
    async getBlogPosts(): Promise<Array<BlogPost>>;
}
```

```ts
//SqliteBlogService.ts
import { IBlogService } from './services/IBlogService';
import { BlogPost } from '../models/BlogPost';

export class SqliteBlogService implements IBlogService {
    async getBlogPosts(): Promise<Array<BlogPost>>{ /* ... */ };
}
```

```ts
//BlogController.ts

//SqliteBlogService @Service is never called because it's never explicitly imported anywhere
import { IBlogService } from '../services/IBlogService';

@Controller({
    route: '/api/blogposts'
})
export class BlogController {
    constructor(private blogService: IBlogService) { }

    @GET()
    public async getBlogPosts(req, res): Promise<Array<BlogPosts>> {
        return this.blogService.getBlogPosts();
    }
}
```

Top down injection could be something like:

```ts
//app.ts
import { injectables } from 'fastify-decorators';
import { IBlogService } from './services/IBlogService';

//injection goes before constructor discovery
if(environment === 'development') {
    injectables.injectSingleton<IBlogService>(new SqliteBlogService());
} else if(environment === 'production') {
    injectables.injectSingleton<IBlogService>(new MySQLBlogService());
}

fastify.register(bootstrap, {
    /* ... */
})
```

this would allow services to be defined as interfaces so that you're not dependent on any particular implementation within the controllers, which is the point of DI

The workaround is to simply import whatever implementations you want in app.ts so that the @Service decorator is called before the @Inject decorators in the controller, but this is obviously code smell

**Comment by L2jLiga** [comment]
Hi there, it's currently near possible with using CJS, for that purpose you have to import `injectables` itself from https://github.com/L2jLiga/fastify-decorators/blob/v3/lib/registry/injectables.ts

Then just set needed value like `injectables.set(IBlogService, SQliteServiceBlog)`

This way you can tell library to use SQliteServiceBlog as IBlogService. Please note that IBlogService shouldn't be decorated as service, but SQliteServiceBlog should.
Hope this helps to workaround it right now, for future it can be considered as one more friendly API

**Comment by L2jLiga** [comment]
> The workaround is to simply import whatever implementations you want in app.ts so that the @service decorator is called before the @Inject decorators in the controller, but this is obviously code smell

It's purely depends on you bootstrap style, in case when you use autoload option and import service directly within you index.ts, then import of service is executed before controllers and their dependencies are loaded, but it can not work in case when you specify controllers to be loaded and it can depends on imports order.
Long story short: it depends who was required/imported first, but yeah, can agree, this is code smell

**Comment by ShortRoundDev** [comment]
@L2jLiga thanks a lot for the help! When I had tried to import `injectables` in my project I got an error that it wasn't exported, but I'll try again, I may have something misconfigured.

**Comment by L2jLiga** [comment]
Could you check this version? `injectables` now exported and has 2 methods accessible via public API - `injectSingleton` to provide singleton object/class instance and so on, and `injectService` to provide class annotated with `@Service`

[fastify-decorators-3.14.1.tgz](https://github.com/L2jLiga/fastify-decorators/files/9939322/fastify-decorators-3.14.1.tgz)

**Comment by L2jLiga** [comment]
Published as [v3.15.0](https://github.com/L2jLiga/fastify-decorators/releases/tag/v3.15.0)

**Comment by ShortRoundDev** [comment]
Wow, great work! Quick progress as well
